crux-core/src/crux/memory.clj

```diff
@@ -227,7 +227,13 @@
              (.putByte idx (unchecked-byte (inc b))))))))))
 
 (defn <-nippy-buffer [buf]
-  (nippy/thaw-from-in! (DataInputStream. (DirectBufferInputStream. buf))))
+  (nippy/thaw-from-in! (-> (DirectBufferInputStream. buf)
+                           (DataInputStream.))))
 
 (defn ->nippy-buffer [v]
-  (->off-heap (nippy/fast-freeze v)))
+  (let [to (ExpandableDirectByteBuffer. 64)
```

Comment:
You probably want to support both on and off heap directly.
The above will allocate off heap, only to copy in back on heap. You can use ByteArrayOutputStream in a similar way and wrap it in a DataOutputStream.

Though often there's already a buffer at hand you can encode into though, value->buffer for example gets handed a `to` buffer. This would be a further optimisation though, as this isn't done today, there's always an extra on heap copy made in fast-freeze.

Reply:
There's also a `putBytes` overload on `MutableDirectBuffer` that can take another buffer (regardless if it's on or off heap), so you could also avoid to do the on-heap step that way.

Reply:
Good shout. I've added an arity to take an output buffer, defaulted to off-heap (because that's what it was previously). For the new usages I've added, I've explicitly specified an on-heap buffer - would that work?

Reply:
Ah, sorry, your second comment didn't show up on the UI until I replied. Will take a look

Reply:
as discussed, try not to copy buffers where possible in the Object value->buffer impl. particularly, id-function can accept a buffer.

Reply:
@hraberg have removed the extra `->on-heap` call. Looking for a way to use `slice-buffer` with an expandable ByteBuffer - we don't currently have a version of `slice-buffer` without a `limit` parameter (and, indeed, there isn't an `UnsafeBuffer` constructor that doesn't require a `limit` parameter) - I've left this one for now. Alternative might be to pass a buffer and an offset into `->nippy-buffer`, but given that's not a pattern we use elsewhere I've not opted for it for now?